File: #0486.predict-the-winner.cpp
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {

        int size = nums.size();
        if (size <= 2) {
            return true;
        }
        
        vector<vector<int>> result = 
            vector<vector<int>>(size, vector<int>(size, -1));
        for (int i = 1; i < size; i++) {
            for (int j = 0; j + i < size; j++) {
                int dist = j + i;
                if (i == 1) {
                    result[i][j] = std::max(nums[j], nums[dist]);
                    continue;
                }

                result[i][j] = std::max(
                    nums[j] + std::accumulate(nums.begin() + j + 1,
                          nums.begin() + dist + 1, 0) - result[i - 1][j + 1],
                    nums[dist] + std::accumulate(nums.begin() + j,
                          nums.begin() + dist, 0) - result[i - 1][j]);
            }
        }
        return result[size - 1][0] >=
            std::accumulate(nums.begin(), nums.end(), 0) / 2.0;
    }
};

File: #0486.predict-the-winner.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var PredictTheWinner = function(nums) {
    let length = nums.length;
    let dp = [];
    for (let i = 0; i < length; i++) {
        dp[i] = new Array(length).fill(0);
        dp[i][i] = nums[i];
    }
    
    for (let i = 2; i <= length; i++) {
        for (let start = 0; start < length - i + 1; start++) {
            const end = start + i - 1;
            dp[start][end] = Math.max(nums[start] - dp[start + 1][end],
                nums[end] - dp[start][end - 1]);
        }
    }
    
    return dp[0][length - 1] >= 0;
};
